src/plugins/interactive_setup/server/routes/enroll.ts

```diff
@@ -6,26 +6,143 @@
 /**
  * Defines routes to deal with Elasticsearch `enroll_kibana` APIs.
  */
-export function defineEnrollRoutes({ router }: RouteDefinitionParams) {
+export function defineEnrollRoutes({
+  router,
+  logger,
+  initializerContext,
+  core,
+  completeSetup,
+}: RouteDefinitionParams) {
   router.post(
     {
       path: '/internal/interactive_setup/enroll',
       validate: {
-        body: schema.object({ token: schema.string() }),
+        body: schema.oneOf([
+          schema.object({
+            hosts: schema.arrayOf(schema.string(), { minSize: 1 }),
+            apiKey: schema.string(),
+            caFingerprint: schema.string(),
+          }),
+          schema.object({
+            hosts: schema.arrayOf(schema.string(), { minSize: 1 }),
+            username: schema.string(),
+            password: schema.string(),
+            caCert: schema.string(),
+          }),
+        ]),
       },
       options: { authRequired: false },
     },
     async (context, request, response) => {
-      return response.forbidden({
-        body: { message: `API is not implemented yet.` },
-      });
+      const client = core.elasticsearch
+        .createClient('enroll', {
+          hosts: request.body.hosts,
+          ssl: { verificationMode: 'none' },
+          // caFingerprint: request.body.caFingerprint,
+        })
+        .asScoped({
+          headers: {
+            authorization:
+              'apiKey' in request.body
+                ? `ApiKey ${btoa(request.body.apiKey)}`
+                : `Basic ${btoa(`${request.body.username}:${request.body.password}`)}`,
+          },
+        });
+
+      const configPath = initializerContext.env.mode.dev
+        ? initializerContext.env.configs.find((fpath) => path.basename(fpath).includes('dev'))
+        : initializerContext.env.configs[0];
+
+      if (!configPath) {
+        logger.error('Cannot find config file');
+        return response.customError({ statusCode: 500, body: 'Cannot find config file.' });
+      }
+      // TODO: Use fingerprint/timestamp as file name
+      const caPath = path.join(path.dirname(configPath), 'ca.crt');
+
+      try {
+        await Promise.all([
+          fs.access(configPath, constants.W_OK),
+          fs.access(path.dirname(caPath), constants.W_OK),
+        ]);
+
+        const { body } = await client.asCurrentUser.transport.request({
+          method: 'GET',
+          path: '/_security/enroll/kibana',
+        });
+
+        const certificateAuthority = generateCertificate(body.http_ca);
+
+        // Ensure we can connect using CA before writing config
+        const verifyConnectionClient = core.elasticsearch.createClient('verifyConnection', {
+          hosts: request.body.hosts,
+          username: 'kibana_system',
+          password: body.password,
+          ssl: { certificateAuthorities: [certificateAuthority] },
+        });
+
+        await verifyConnectionClient.asInternalUser.security.authenticate();
+
+        await Promise.all([
```

Comment:
__note:__ I'd probably make these calls sequentially, not a big deal performance wise, but it'd allow to be 100% sure that CA certificate file is successfully persisted before we alter configuration file.